ch_util/tools.py

```diff
@@ -2354,5 +2354,23 @@ def invert_no_zero(x):
     r : np.ndarray
         Return the reciprocal of x.
     """
-    with np.errstate(divide="ignore", invalid="ignore"):
-        return np.where(x == 0, 0.0, 1.0 / x)
+    return np.divide(1, x, out=np.zeros_like(x), where=x != 0)
```

Comment:
Weirdly, I've already fixed this up somewhere, but the commit must have been lost in all of the merging and rebasing we've been doing.

Ideally there should only be one `invert_no_zero` routine and it should support both arguments. I think you can do that via:
```suggestion
    out = np.zeros_like(x)
    np.divide(1, x, out=out.view(np.ndarray), where=x != 0)
    return out
```
and then just remove the `_mpi` version.

Also note that there is a much faster implementation of this available in `draco` at (https://github.com/radiocosmology/draco/blob/master/draco/util/tools.py#L226). The problem with using it is that `ch_util` doesn't have a dependency on `draco` at the moment.

Reply:
Interesting about the draco one, althoughI understand the dependency issue. At the moment, this won't work for mpiarrays after the new updates because "where" operations aren't supported. In`np.divide`, it tries to call `where` in all 3 spots that x exists, so we can only really pass `np.ndarray` into it, unless we update

Reply:
We could merge the functionality into a single function which checks for type first though

Reply:
I think it would work actually fine, the `.view` call is effectively a `.local_array` call, but one which works on `np.ndarray`'s too.

Reply:
Oh sorry. You also need an `x.view(np.ndarray)`

Reply:
Do you mean passing the local array to the function only? If I run just those lines where x is a `MPIArray` it gives a `UnsupportedOperation` error

Reply:
i.e. this is fine
```python
from caput import mpiarray
import numpy as np

x = mpiarray.MPIArray((4, 4), dtype=np.float64)
x[:] = 3

out = np.zeros_like(x)

np.divide(1, x.view(np.ndarray), out=out.view(np.ndarray), where=a != 0)
```